library/init/category/lawful.lean

```diff
@@ -0,0 +1,223 @@
+attribute [simp] id_map
+-- `comp_map` does not make a good simp lemma
+
+class is_lawful_applicative (f : Type u → Type v) [applicative f] extends is_lawful_functor f : Prop :=
+(seq_left_eq  : ∀ {α β : Type u} (a : f α) (b : f β), a <* b = const β <$> a <*> b . control_laws_tac)
+(seq_right_eq : ∀ {α β : Type u} (a : f α) (b : f β), a *> b = const α id <$> a <*> b . control_laws_tac)
+-- applicative laws
+(pure_seq_eq_map : ∀ {α β : Type u} (g : α → β) (x : f α), pure g <*> x = g <$> x) -- . control_laws_tac)
+(map_pure        : ∀ {α β : Type u} (g : α → β) (x : α), g <$> (pure x : f α) = pure (g x))
+(seq_pure        : ∀ {α β : Type u} (g : f (α → β)) (x : α), g <*> pure x = (λ g : α → β, g x) <$> g)
+(seq_assoc       : ∀ {α β γ : Type u} (x : f α) (g : f (α → β)) (h : f (β → γ)), h <*> (g <*> x) = (@comp α β γ <$> h) <*> g <*> x)
+-- default functor law
+(comp_map := begin intros; simp [(pure_seq_eq_map _ _).symm, seq_assoc, map_pure, seq_pure] end)
+
+export is_lawful_applicative (seq_left_eq seq_right_eq pure_seq_eq_map map_pure seq_pure seq_assoc)
+attribute [simp] map_pure seq_pure
+
+-- applicative "law" derivable from other laws
+@[simp] theorem pure_id_seq {α : Type u} {f : Type u → Type v} [applicative f] [is_lawful_applicative f] (x : f α) : pure id <*> x = x :=
+by simp [pure_seq_eq_map]
+
+class is_lawful_monad (m : Type u → Type v) [monad m] extends is_lawful_applicative m : Prop :=
+(bind_pure_comp_eq_map : ∀ {α β : Type u} (f : α → β) (x : m α), x >>= pure ∘ f = f <$> x  . control_laws_tac)
+(bind_map_eq_seq : ∀ {α β : Type u} (f : m (α → β)) (x : m α), f >>= (<$> x) = f <*> x  . control_laws_tac)
+-- monad laws
+(pure_bind : ∀ {α β : Type u} (x : α) (f : α → m β), pure x >>= f = f x)
+(bind_assoc : ∀ {α β γ : Type u} (x : m α) (f : α → m β) (g : β → m γ),
+  x >>= f >>= g = x >>= λ x, f x >>= g)
+-- helper lemma
+(bind_bind_pure_comp_eq_seq : ∀ {α β : Type u} (f : m (α → β)) (x : m α), (f >>= λ f, x >>= pure ∘ f) = f <*> x :=
+  by intros; rw ←bind_map_eq_seq; simp [(bind_pure_comp_eq_map _ _).symm])
+-- all applicative laws are derivable from the monad laws + id_map
+(pure_seq_eq_map := by intros; rw ←bind_map_eq_seq; simp [pure_bind])
+(map_pure := by intros; rw ←bind_pure_comp_eq_map; simp [pure_bind])
+(seq_pure := by intros; rw ←bind_map_eq_seq; simp [map_pure, bind_pure_comp_eq_map])
+(seq_assoc := by intros; simp [(bind_pure_comp_eq_map _ _).symm,
+                                (bind_bind_pure_comp_eq_seq _ _).symm, bind_assoc, pure_bind])
+
+export is_lawful_monad (bind_pure_comp_eq_map bind_bind_pure_comp_eq_seq bind_map_eq_seq pure_bind bind_assoc)
+attribute [simp] pure_bind
+
+-- monad "law" derivable from other laws
+@[simp] theorem bind_pure {α : Type u} {m : Type u → Type v} [monad m] [is_lawful_monad m] (x : m α) : x >>= pure = x :=
+show x >>= pure ∘ id = x, by rw bind_pure_comp_eq_map; simp [id_map]
+
+lemma bind_ext_congr {α β} {m : Type u → Type v} [has_bind m] {x : m α} {f : α → m β} (g : α → m β) :
```

Comment:
Why is `f` implicit and `g` explicit?

Reply:
In principle, both can be inferred from `(∀ a, f a = g a)`

Reply:
It was useful in the original commit, but now it doesn't matter anymore. Changed.

Reply:
Btw, I will append new commits for changes for now and squash them in the end.